src/aspire/orientation/commonline_sync.py

```diff
@@ -1,31 +1,443 @@
+            # Note that as long as the diagonal of the common lines matrix is
+            # zero, the conditions (i != j) && (j != k) are not needed, since
+            # if i == j then clmatrix[i, k] == 0 and similarly for i == k or
+            # j == k. Thus, the previous voting code (from the JSB paper) is
+            # correct even though it seems that we should test also that
+            # (i != j) && (i != k) && (j != k), and only (i != j) && (i != k)
+            #  as tested there.
+
+            l_idx12 = clmatrix[i, j]
+            l_idx21 = clmatrix[j, i]
+            k_list = k_list[np.logical_and(
+                np.logical_and(k_list != i, clmatrix[i, k_list] != -1), clmatrix[j, k_list] != -1)]
+
+            l_idx13 = clmatrix[i, k_list]
+            l_idx31 = clmatrix[k_list, i]
+            l_idx23 = clmatrix[j, k_list]
+            l_idx32 = clmatrix[k_list, j]
+
+            # theta1 is the angle on C1 created by its intersection with C3 and C2.
+            # theta2 is the angle on C2 created by its intersection with C1 and C3.
+            # theta3 is the angle on C3 created by its intersection with C2 and C1.
+            theta1 = (l_idx13 - l_idx12) * 2 * np.pi / n_theta
+            theta2 = (l_idx21 - l_idx23) * 2 * np.pi / n_theta
+            theta3 = (l_idx32 - l_idx31) * 2 * np.pi / n_theta
+
+            c1 = np.cos(theta1)
+            c2 = np.cos(theta2)
+            c3 = np.cos(theta3)
+
+            # Each common-line corresponds to a point on the unit sphere. Denote the
+            # coordinates of these points by (Pix, Piy Piz), and put them in the matrix
+            #   M=[ P1x  P2x  P3x ;
+            #       P1y  P2y  P3y ;
+            #       P1z  P2z  P3z ].
+            #
+            # Then the matrix
+            #   C=[  1  c1  c2 ;
+            #       c1   1  c3 ;
+            #       c2  c3   1 ],
+            # where c1,c2,c3 are given above, is given by C = M.T @ M.
+            # For the points P1,P2, and P3 to form a triangle on the unit sphere, a
+            # necessary and sufficient condition is for C to be positive definite. This
+            # is equivalent to
+            #       1+2*c1*c2*c3-(c1^2+c2^2+c3^2) > 0.
+            # However, this may result in a triangle that is too flat, that is, the
+            # angle between the projections is very close to zero. We therefore use the
+            # condition below
+            #       1+2*c1*c2*c3-(c1^2+c2^2+c3^2) > 1.0e-5.
+            # This ensures that the smallest singular value (which is actually
+            # controlled by the determinant of C) is big enough, so the matrix is far
+            # from singular. This condition is equivalent to computing the singular
+            # values of C, followed by checking that the smallest one is big enough.
+            cond = 1 + 2 * c1 * c2 * c3 - (
+                    np.square(c1) + np.square(c2) + np.square(c3))
+
+            good_idx = np.where(cond > 1e-5)[0]
```

Comment:
Yes. The same value in Matlab version.

Reply:
No. I'm saying that you don't need the `np.where` (I think this is just a translation of Matlab's `find` but it shouldn't be necessary).

Reply:
Remove np.where can not pass unit test. For the code in `_vote_ij` function, the good_idx need to store the good indices only instead of a mask with the full length including false values to bad ones.

Reply:
It's not clear to my why that's the case, but if we need to get the indices, I think the canonical way of doing this is using `np.nonzero` (the `np.where` function is used to select elements between two arrays depending on some condition).

Reply:
Switched to `np.nonzero`.